File: Cassandra.Client/CassandraClient.cpp
#include "CassandraClient.h"

namespace Cassandra
{
    namespace Client
    {
        bool TransportKeyspaceIsNotSet(CassandraTransport^ transport, IArgs^ args)
        {
            return args->Keyspace != nullptr && transport->_keyspace != args->Keyspace;
        }

        void NotifyCompleted(uv_async_t* notifier, int status)
        {
            CassandraContextQueue^ contextQueue = CassandraContextQueue::FromPointer(notifier->data);
            CassandraContext^ context;

            while (contextQueue->TryDequeue(context))
            {
                IPEndPoint^ endPoint = context->_args->EndPoint;

                CassandraClient^ client = context->_client;

                // try get transport from a pool
                CassandraTransport^ transport = (CassandraTransport^)client->_transportPool->Get(endPoint);

                // or try create new if no pooled transports are available
                if (transport == nullptr)
                {
                    transport = client->_factory->CreateTransport(endPoint);

                    // transport limit reached
                    if (transport == nullptr)
                    {
                        context->_resultCallback(nullptr, gcnew TTransportException("Transport limit reached."));
                        continue;
                    }
                }

                // check if keyspace needs to be set
                if (TransportKeyspaceIsNotSet(transport, context->_args))
                {
                    // send set keyspace args before the original args is sent

                    String^ keyspace = context->_args->Keyspace;
                    SetKeyspaceArgs^ setKeyspaceArgs = gcnew SetKeyspaceArgs(endPoint, keyspace);
                    SetKeyspaceContext^ setKeyspaceContext = gcnew SetKeyspaceContext(context, keyspace);
                    ResultCallback^ setKeyspaceCompleted = gcnew ResultCallback(setKeyspaceContext, &SetKeyspaceContext::Completed);
                    transport->_context = gcnew CassandraContext(setKeyspaceArgs, setKeyspaceCompleted, client);
                }
                else
                {
                    transport->_context = context;
                }

                transport->_context->_args->WriteMessage(transport->_protocol);
            }
        }


        void StopCompleted(uv_async_t* stop, int status)
        {
            uv_stop(stop->loop);
        }


        CassandraClient::CassandraClient(int maxEndPointTransportCount)
        {
            _loop = uv_loop_new();

            _contextQueue = gcnew CassandraContextQueue();
            _transportPool = gcnew CassandraTransportPool();
            _factory = gcnew CassandraTransport::Factory(maxEndPointTransportCount, _loop);

            _notifier = new uv_async_t();
            _notifier->data = _contextQueue->ToPointer();
            uv_async_init(_loop, _notifier, NotifyCompleted);

            _stop = new uv_async_t();
            uv_async_init(_loop, _stop, StopCompleted);
        }


        CassandraClient::~CassandraClient()
        {
            uv_loop_delete(_loop);

            delete _notifier;
            delete _stop;
            delete _contextQueue;

            for each(IPEndPoint^ endPoint in _transportPool->GetEndPoints())
            {
                while (true)
                {
                    ICassandraTransport^ transport = _transportPool->Get(endPoint);

                    if (transport == nullptr)
                    {
                        break;
                    }

                    transport->Close();
                }
            }
        }


        void CassandraClient::Stop()
        {
            uv_async_send(_stop);
        }


        void CassandraClient::Send(IArgs^ args, ResultCallback^ resultCallback)
        {
            if (args == nullptr)
                throw gcnew ArgumentNullException("args");

            if (resultCallback == nullptr)
                throw gcnew ArgumentNullException("resultCallback");

            CassandraContext^ context = gcnew CassandraContext(args, resultCallback, this);

            _contextQueue->Enqueue(context);

            // notify that there are requests pending
            uv_async_send(_notifier);
        }


        void CassandraClient::Run()
        {
            uv_run(_loop, UV_RUN_DEFAULT);
        }


        CassandraClient^ CassandraClient::RunAsync()
        {
            Thread^ loopThread = gcnew Thread(gcnew ThreadStart(this, &CassandraClient::RunInternal));
            loopThread->IsBackground = true;
            loopThread->Start();
            return this;
        }


        void CassandraClient::RunInternal()
        {
            this->Run();
            delete this;
        }
    }
}

File: Cassandra.Client/CassandraClient.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Net;
using System.Threading;
using Cassandra.Client.Thrift;
using NetUv;
using log4net;

namespace Cassandra.Client
{
    public sealed class CassandraClient : IDisposable
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(CassandraClient));

        private readonly IUvLoop _loop;
        private readonly ITransportFactory _transportFactory;
        private readonly IClientStats _clientStats;

        // wait handle for signaling loop stop
        private readonly ManualResetEventSlim _loopStop;

        // uv handle for stopping the loop
        private readonly IUvAsync _asyncStop;

        // uv handle for notifying the loop about new requests
        private readonly IUvAsync _asyncSend;

        private readonly ConcurrentQueue<CassandraContext> _contextQueue;

        private readonly ITransportPool _transportPool;

        public CassandraClient()
            : this(new UvFramedTransport.Factory(DefaultClientStats),
                DefaultClientStats)
        {
        }

        public CassandraClient(IClientStats clientStats)
            : this(new UvFramedTransport.Factory(clientStats), clientStats)
        {
        }

        public CassandraClient(ITransportFactory transportFactory,
            IClientStats clientstats)
            : this(UvLoop.Default, transportFactory, clientstats)
        {
        }

        internal CassandraClient(IUvLoop loop,
            ITransportFactory transportFactory,
            IClientStats clientStats)
        {
            _loop = loop;
            _transportFactory = transportFactory;
            _clientStats = clientStats;

            _loopStop = new ManualResetEventSlim();
            _contextQueue = new ConcurrentQueue<CassandraContext>();
            _transportPool = new CassandraTransportPool();

            // close all active loop handles, so loop thread can return
            _asyncStop = _loop.InitUvAsync((async, exception) => CloseAllHandles());

            // notify context queue
            _asyncSend = _loop.InitUvAsync((async, exception) => ProcessContextQueue());
        }

        public void SendAsync(IArgs args, ResultCb resultCb)
        {
            var context = new CassandraContext(args, resultCb);

            // enqueue and notify the loop about contexts pending
            _contextQueue.Enqueue(context);
            _asyncSend.Send();
            _clientStats.IncrementArgsEnqueued();
        }

        public void RunAsync()
        {
            Debug.WriteLine("Main thread ID={0}", Thread.CurrentThread.ManagedThreadId);

            new Thread(() =>
                {
                    Debug.WriteLine("Loop thread ID={0}", Thread.CurrentThread.ManagedThreadId);

                    try
                    {
                        _loop.Run();
                    }
                    catch (Exception e)
                    {
                        Log.Error("Loop thread error.", e);
                    }

                    _loopStop.Set();
                })
                {
                    IsBackground = true
                }
            .Start();
        }

        public void Dispose()
        {
            Debug.WriteLine("Disposing on thread ID={0}", Thread.CurrentThread.ManagedThreadId);

            // notify loop to stop
            _asyncStop.Send();

            // wait until loop stops
            _loopStop.Wait();

            // it is not nice to dispose the default loop
            if (!ReferenceEquals(_loop, UvLoop.Default))
            {
                // check if loop implements disposable
                var loop = _loop as IDisposable;

                if (loop != null)
                {
                    loop.Dispose();
                }
            }

            // dispose wait handle
            _loopStop.Dispose();
        }

        private void ProcessContextQueue()
        {
            Debug.WriteLine("ProcessContextQueue on thread ID={0}", Thread.CurrentThread.ManagedThreadId);

            CassandraContext context;

            while (_contextQueue.TryDequeue(out context))
            {
                try
                {
                    ProcessContext(context);
                }
                finally
                {
                    _clientStats.IncrementArgsDequeued();
                }
            }
        }

        private void ProcessContext(CassandraContext context)
        {
            var endPoint = context.Args.EndPoint;

            ITransport transport;
            if (!_transportPool.TryGet(endPoint, out transport))
            {
                if (!_transportFactory.TryCreate(endPoint, out transport))
                {
                    context.ResultCb(null,
                        new TransportLimitException("Transport limit reached.",
                            endPoint));
                    return;
                }
            }

            var wrappedContext = GetKeyspaceWrappedContext(transport, context);
            wrappedContext.SendArgs(transport);
        }

        private CassandraContext GetKeyspaceWrappedContext(ITransport transport, CassandraContext context)
        {
            if (TransportKeyspaceIsNotSet(transport, context.Args))
            {
                var keyspace = context.Args.Keyspace;
                var setKeyspaceArgs = new SetKeyspaceArgs(transport.EndPoint, keyspace);

                return new CassandraContext(setKeyspaceArgs,
                    (trans, exception) =>
                        {
                            if (exception != null)
                            {
                                context.ResultCb(trans, exception);
                                return;
                            }

                            transport.Keyspace = keyspace;
                            context.TransportPool = _transportPool; // return to pool only after keyspace is set
                            context.SendArgs(trans);
                        });
            }

            context.TransportPool = _transportPool;
            return context;
        }

        private static bool TransportKeyspaceIsNotSet(ITransport transport, IArgs args)
        {
            return args.Keyspace != null && transport.Keyspace != args.Keyspace;
        }

        private void CloseAllHandles()
        {
            Debug.WriteLine("CloseAllHandles on thread ID={0}", Thread.CurrentThread.ManagedThreadId);

            _asyncSend.Close(h => h.Dispose());
            _asyncStop.Close(h => h.Dispose());
        }

        private static readonly IClientStats DefaultClientStats = new ClientStats();

        private class ClientStats : IClientStats
        {
            public void IncrementTransportOpen(EndPoint endPoint)
            {
            }

            public void IncrementTransportClose(EndPoint endPoint)
            {
            }

            public void IncrementTransportSendFrame(EndPoint endPoint)
            {
            }

            public void IncrementTransportReceiveFrame(EndPoint endPoint)
            {
            }

            public void IncrementArgsEnqueued()
            {
            }

            public void IncrementArgsDequeued()
            {
            }
        }
    }
}
